nextcore/endpoint/interactions.py

```diff
@@ -0,0 +1,158 @@
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+from __future__ import annotations
+
+import asyncio
+
+from aiohttp import web
+from json import JSONDecodeError
+from logging import getLogger
+from nacl.exceptions import BadSignatureError
+from nacl.signing import VerifyKey
+from typing import TYPE_CHECKING
+
+from nextcore.common import Dispatcher
+
+
+if TYPE_CHECKING:
+    from typing import Final
+
+    from discord_typings import InteractionData
+    from multidict import CIMultiDictProxy
+
+
+logger = getLogger(__name__)
+
+
+__all__: Final[tuple[str, ...]] = (
+    "InteractionEndpoint",
+    "verify_signature",
+)
+
+
+def verify_signature(public_key: str, signature: str, timestamp: str, body: str) -> bool:
+    verify_key = VerifyKey(bytes.fromhex(public_key))
+    verify_key.verify(f"{timestamp}{body}".encode(), bytes.fromhex(signature))
+    return True
+
+
+class InteractionEndpoint:
+    _public_key: str | None
+    _route: str | None
+    event_dispatcher: Dispatcher
+
+    # def __init__(
+    #         self,
+    #         *,
+    #         web_app: web.Application = None,
+    #         route_path: str = "/interactions/test"
+    # ):
+    #     self.public_key: str | None = None
+    #     self.app: web.Application = web_app or web.Application(middlewares=[interaction_endpoint("asdf", "/interactions/test")])
+    #     self.site: web.TCPSite | None = None
+    #     self.event_dispatcher: Dispatcher = Dispatcher()
+    #
+    #     self.app.add_routes([
+    #         web.route("POST", route_path, self._on_interaction_incoming)
+    #     ])
+
+    def __init__(self):
+        self._public_key = None
+        self._route = None
+        self.event_dispatcher = Dispatcher()
+
+    def _verify_from_headers(self, headers: CIMultiDictProxy, body: str) -> bool:
+        try:
+            signature: str = headers.get("x-signature-ed25519")
+            timestamp: str = headers.get("x-signature-timestamp")
+            verify_signature(self._public_key, signature, timestamp, body)
+            return True
+        except BadSignatureError:
+            return False
+
+    async def _on_interaction_incoming(self, request: web.Request) -> web.Response | None:
+        logger.debug("Incoming interaction endpoint request.")
+        try:
+            body = (await request.read()).decode("utf-8")
+            if self._verify_from_headers(request.headers, body):
+                logger.debug("Verified Discord signature headers, continuing.")
+                interaction: InteractionData = await request.json()
+                if interaction.get("type") == 1:
+                    # logger.debug("Dispatching ping.")
+                    await self.event_dispatcher.dispatch("ping", request, interaction)
+                else:
+                    await self.event_dispatcher.dispatch("interaction", request, interaction)
+
+                return None
```

Comment:
Isn't this a little unnecessary? If the code gets to this point, the only thing it can do is just return.